src/vsr/replica.zig

```diff
@@ -6824,6 +7427,311 @@ pub fn ReplicaType(
+                    self.replica,
+                    @tagName(state_old),
+                    @tagName(self.sync_stage),
+                    target.checkpoint_op,
+                    target.checkpoint_id,
+                });
+            } else {
+                log.debug("{}: sync_dispatch: {s}..{s}", .{
+                    self.replica,
+                    @tagName(state_old),
+                    @tagName(self.sync_stage),
+                });
+            }
+
+            if (self.test_event_callback) |hook| hook(self, .sync);
+
+            switch (self.sync_stage) {
+                .not_syncing => {},
+                .cancelling_commit => {}, // Waiting for an uninterruptible commit step.
+                .cancelling_grid => {
+                    self.grid_repair_message_timeout.stop();
+                    self.grid.cancel(sync_cancel_grid_callback);
+
+                    assert(self.grid.read_faulty_queue.empty());
+                },
+                .requesting_target => {}, // Waiting for a usable sync target.
+                .request_trailers => self.sync_message_timeout.start(),
+                .updating_superblock => self.sync_superblock_update(),
+            }
+        }
+
+        fn sync_cancel_commit_callback(self: *Self) void {
+            assert(!self.solo());
+            assert(self.sync_stage == .cancelling_commit);
+
+            switch (self.commit_stage) {
+                .idle,
+                .next,
+                .next_pipeline,
+                .cleanup,
+                .prefetch_state_machine,
+                .compact_state_machine,
+                .checkpoint_state_machine,
+                => unreachable,
+
+                .next_journal,
+                .setup_client_replies,
+                .checkpoint_client_replies,
+                .checkpoint_superblock,
+                => {},
+            }
+
+            // Even though the commit flow has been stopped, there may still be pending IO
+            // between beats.
+            // TODO Compaction pacing: If there is no grid IO between beats, we can transition
+            // to stage=requesting_target.
+            self.sync_dispatch(.cancelling_grid);
+        }
+
+        fn sync_cancel_grid_callback(grid: *Grid) void {
+            const self = @fieldParentPtr(Self, "grid", grid);
+            assert(self.sync_stage == .cancelling_grid);
+            assert(self.grid.read_queue.empty());
+            assert(self.grid.read_faulty_queue.empty());
+
+            if (self.commit_stage == .idle) {
+                assert(self.commit_prepare == null);
+            } else {
+                if (self.commit_prepare) |prepare| self.message_bus.unref(prepare);
+                self.commit_prepare = null;
+                self.commit_stage = .idle;
+            }
+
+            var grid_reads = self.grid_reads.iterate();
+            while (grid_reads.next()) |grid_read| {
+                self.message_bus.unref(grid_read.message);
```

Comment:
Do we have any further expectations w.r.t. the number of references in `grid_read.message` now?